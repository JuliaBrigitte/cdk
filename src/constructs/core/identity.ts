export interface Identity {
  stack: string;
  stage: string;
  app: string;
}
