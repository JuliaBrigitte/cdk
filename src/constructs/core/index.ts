export * from "./identity";
export * from "./mappings";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
