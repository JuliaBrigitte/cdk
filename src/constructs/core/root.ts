import type { StageSynthesisOptions } from "aws-cdk-lib";
import { App } from "aws-cdk-lib";
import type { CloudAssembly } from "aws-cdk-lib/cx-api";
import { RiffRaffYamlFile } from "../../utils/riff-raff-yaml-file";

// Deliberately not called `GuApp` as `app` is overloaded (as is `stack`, and `stage` ðŸ˜…)
export class GuRoot extends App {
  override synth(options?: StageSynthesisOptions): CloudAssembly {
    new RiffRaffYamlFile(this).synth();
    return super.synth(options);
  }
}
