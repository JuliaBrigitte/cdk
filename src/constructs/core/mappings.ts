import { CfnMapping } from "@aws-cdk/core";
import type { GuStack } from "./stack";

export interface GuStageDependentValue {
  variableName: string;
  codeValue: unknown;
  prodValue: unknown;
}

export class GuStageMapping extends CfnMapping {
  constructor(scope: GuStack, id: string = "stage-mapping") {
    super(scope, id);
  }
}
