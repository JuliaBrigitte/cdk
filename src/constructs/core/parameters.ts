import type { CfnParameterProps } from "@aws-cdk/core";
import { CfnParameter } from "@aws-cdk/core";
import { Stage, Stages } from "../../constants";
import type { GuStack } from "./stack";

export type GuParameterProps = CfnParameterProps;

export type GuNoTypeParameterProps = Omit<GuParameterProps, "type">;

export class GuParameter extends CfnParameter {
  constructor(scope: GuStack, id: string, props: GuParameterProps) {
    super(scope, id, props);
  }
}

export class GuStringParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "String" });
  }
}

export class GuStageParameter extends GuParameter {
  constructor(scope: GuStack) {
    super(scope, "Stage", {
      type: "String",
      description: "Stage name",
      allowedValues: Stages,
      default: Stage.CODE,
    });
  }
}

export class GuStackParameter extends GuParameter {
  constructor(scope: GuStack) {
    super(scope, "Stack", {
      type: "String",
      description: "Name of this stack",
      default: "deploy",
    });
  }
}

// TODO: Is there a way of removing default props if they weren't implemented before?
//       Should that be allowed even if it is possible?
export class GuInstanceTypeParameter extends GuParameter {
  constructor(scope: GuStack, id: string = "InstanceType", props: GuParameterProps = {}) {
    super(scope, id, {
      type: "String",
      description: "EC2 Instance Type",
      default: "t3.small",
      ...props,
    });
  }
}

export class GuSSMParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      noEcho: true,
      ...props,
      type: "AWS::SSM::Parameter::Value<String>",
    });
  }
}

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

export class GuVpcParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}

export class GuAmiParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::Image::Id",
    });
  }
}
