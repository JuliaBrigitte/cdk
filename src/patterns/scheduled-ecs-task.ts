import { Alarm, TreatMissingData } from "@aws-cdk/aws-cloudwatch";
import { SnsAction } from "@aws-cdk/aws-cloudwatch-actions";
import type { ISecurityGroup, IVpc } from "@aws-cdk/aws-ec2";
import type { IRepository } from "@aws-cdk/aws-ecr";
import {
  Cluster,
  Compatibility,
  ContainerImage,
  FargatePlatformVersion,
  LogDrivers,
  TaskDefinition,
} from "@aws-cdk/aws-ecs";
import type { Schedule } from "@aws-cdk/aws-events";
import { Rule } from "@aws-cdk/aws-events";
import { SfnStateMachine } from "@aws-cdk/aws-events-targets";
import type { PolicyStatement } from "@aws-cdk/aws-iam";
import { Topic } from "@aws-cdk/aws-sns";
import { IntegrationPattern, StateMachine } from "@aws-cdk/aws-stepfunctions";
import { EcsFargateLaunchTarget, EcsRunTask } from "@aws-cdk/aws-stepfunctions-tasks";
import { CfnOutput, Duration } from "@aws-cdk/core";
import { GuDistributionBucketParameter } from "../constructs/core";
import type { GuStack } from "../constructs/core";
import type { Identity } from "../constructs/core/identity";
import { GuGetDistributablePolicyStatement } from "../constructs/iam";

/**
 * Configuration to determine what container to use for the task
 * For example, to run the task in the latest node container:
 *
 * ```typescript
 * const containerConfiguration = {
 *     id: 'node:latest'
 * }
 * ```
 *
 * Alternatively you can specify a repository and version, for example for an ECR repository:
 * ```typescript
 * import { Repository } from "@aws-cdk/aws-ecr";
 * const repository = new Repository(scope, `${app}-repository`, {
    repositoryName: app,
  });
 * const containerConfiguration = {
 *     repository: Repository.fromRepositoryArn("<repository arn>"),
 *     version: '1'
 * }
 * ```
 */

type RepositoryContainer = {
  repository: IRepository;
  version: string;
  type: "repository";
};

type RegistryContainer = {
  id?: string;
  type: "registry";
};

type ContainerConfiguration = RepositoryContainer | RegistryContainer;

/**
 * Configuration options for the [[`GuScheduledEcsTask`]] pattern.
 *
 * The `schedule` property determines when your task is invoked. For example, to invoke
 * the lambda every 5 minutes, use:
 * ```typescript
 * import { Schedule } from "@aws-cdk/aws-events";
 * import { Duration } from "@aws-cdk/core";
 *
 * const props = {
 *   // Other props here
 *   schedule: Schedule.rate(Duration.minutes(5)),
 * }
 * ```
 *
 * To invoke the task every weekday at 8am, use:
 * ```
 * import { Schedule } from "@aws-cdk/aws-events";
 *
 * const props = {
 *   // Other props here
 *   schedule: Schedule.expression("cron(0 8 ? * MON-FRI *)"),
 * }
 * ```
 *
 * See [[`ContainerConfiguration`]] for details of how to configure the container used to run the task.
 *
 * `taskTimeoutInMinutes` does what is says on the tin. The default timeout is 15 minutes.
 *
 * The `taskCommand` prop allows you to specify a command to run when the task starts (this can also be done via a CMD statement in
 * your Dockerfile). For example:
 *
 * const props = {
 *    //other props
 *    taskCommand: `aws s3 cp s3://${distbucket}/${stack}/${stage}/${app}/task.sh . && ./task.sh
    }
 *
 * It is advisable to configure alarms for when the job fails/times out. To do this specify the alarmSnsTopicArn prop.
 *
 * `customTaskPolicies` allows your task to interact with other AWS services. By default a task
 * will have read access to the distribution bucket for your account.
 *
 * You can specify security groups to apply to task using the `securityGroups` prop.
 *
 * You can also set the memory and cpu units for your task. See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_size
 * for further details.
 *
 */
export interface GuScheduledEcsTaskProps extends Identity {
  vpc: IVpc;
  schedule: Schedule;
  containerConfiguration: ContainerConfiguration;
  taskTimeoutInMinutes?: number;
  cpu?: number;
  memory?: number;
  taskCommand?: string;
  alarmSnsTopicArn?: string;
  securityGroups?: ISecurityGroup[];
  customTaskPolicies?: PolicyStatement[];
}

/**
 * Containers can be specified either via a repository and version or by container id (in which case it will
 * be fetched from docker hub).
 */
const getContainer = (config: ContainerConfiguration) => {
  if (config.type === "repository") {
    return ContainerImage.fromEcrRepository(config.repository, config.version);
  } else {
    return ContainerImage.fromRegistry(config.id ?? "ubuntu:focal");
  }
};

/**
 * Pattern which creates all of the resources needed to invoke a fargate task on a schedule.
 *
 * The task will be wrapped in a step function to allow for easier triggering and alarming on failure.
 *
 * For all configuration options, see [[`GuScheduledEcsTaskProps`]].
 *
 * Note that if your task reliably completes in less than 15 minutes then you should probably use a [[`GuScheduledLambda`]] instead. This
 * pattern was mainly created to work around the 15 minute lambda timeout.
 */
export class GuScheduledEcsTask {
  constructor(scope: GuStack, props: GuScheduledEcsTaskProps) {
    const timeout = props.taskTimeoutInMinutes ?? 15;

    const cpu = props.cpu ?? 2048;
    const memory = props.memory ?? 4096;

    const cluster = new Cluster(scope, `${props.app}-cluster`, {
      clusterName: `${props.app}-cluster-${props.stage}`,
      enableFargateCapacityProviders: true,
      vpc: props.vpc,
    });

    const taskDefinition = new TaskDefinition(scope, `${props.app}-task-definition`, {
      compatibility: Compatibility.FARGATE,
      cpu: cpu.toString(),
      memoryMiB: memory.toString(),
    });

    taskDefinition.addContainer("scheduled-task-container", {
      image: getContainer(props.containerConfiguration),
      entryPoint: props.taskCommand ? ["/bin/sh"] : undefined,
      command: props.taskCommand ? ["-c", `${props.taskCommand}`] : undefined,
      cpu,
      memoryLimitMiB: props.memory,
      logging: LogDrivers.awsLogs({
        streamPrefix: props.app,
        logRetention: 14,
      }),
    });

    const distPolicy = new GuGetDistributablePolicyStatement(scope, { app: props.app });

    taskDefinition.addToTaskRolePolicy(distPolicy);
    (props.customTaskPolicies ?? []).forEach((p) => taskDefinition.addToTaskRolePolicy(p));

    const task = new EcsRunTask(scope, `${props.app}-task`, {
      cluster: cluster,
      launchTarget: new EcsFargateLaunchTarget({
        platformVersion: FargatePlatformVersion.LATEST,
      }),
      taskDefinition: taskDefinition,
      integrationPattern: IntegrationPattern.RUN_JOB,
      resultPath: "DISCARD",
      timeout: Duration.minutes(timeout),
      securityGroups: props.securityGroups ?? [],
    });

    const stateMachine = new StateMachine(scope, `${props.app}-state-machine`, {
      definition: task,
      stateMachineName: `${props.app}-${props.stage}`,
    });

    new Rule(scope, `${props.app}-schedule-rule`, {
      schedule: props.schedule,
      targets: [new SfnStateMachine(stateMachine)],
    });

    if (props.alarmSnsTopicArn) {
      const alarmTopic = Topic.fromTopicArn(scope, `${props.app}-alarm-topic`, props.alarmSnsTopicArn);
      const alarms = [
        {
          name: "ExecutionsFailedAlarm",
          description: `${props.app}-${props.stage} job failed `,
          metric: stateMachine.metricFailed({
            period: Duration.hours(1),
            statistic: "sum",
          }),
        },
        {
          name: "TimeoutAlarm",
          description: `${props.app}-${props.stage} job timed out `,
          metric: stateMachine.metricTimedOut({
            period: Duration.hours(1),
            statistic: "sum",
          }),
        },
      ];

      alarms.forEach((a) => {
        const alarm = new Alarm(scope, a.name, {
          alarmDescription: a.description,
          actionsEnabled: true,
          metric: a.metric,
          // default for comparisonOperator is GreaterThanOrEqualToThreshold
          threshold: 1,
          evaluationPeriods: 1,
          treatMissingData: TreatMissingData.NOT_BREACHING,
        });
        alarm.addAlarmAction(new SnsAction(alarmTopic));
      });
    }

    new CfnOutput(scope, "StateMachineArn", {
      value: stateMachine.stateMachineArn,
    });
  }
}
