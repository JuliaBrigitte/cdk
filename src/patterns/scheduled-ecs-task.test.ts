import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { SecurityGroup, Vpc } from "@aws-cdk/aws-ec2";
import { Schedule } from "@aws-cdk/aws-events";
import { Effect, PolicyStatement } from "@aws-cdk/aws-iam";
import { Duration } from "@aws-cdk/core";
import type { GuStack } from "../constructs/core";
import { simpleGuStackForTesting } from "../utils/test";
import { GuScheduledEcsTask } from "./scheduled-ecs-task";

const vpc = (stack: GuStack) =>
  Vpc.fromVpcAttributes(stack, "VPC", {
    vpcId: "test",
    availabilityZones: [""],
    publicSubnetIds: [""],
    privateSubnetIds: ["abc-123"],
  });

const securityGroup = (stack: GuStack) => SecurityGroup.fromSecurityGroupId(stack, "hehe", "id-123");
const testPolicy = new PolicyStatement({
  effect: Effect.ALLOW,
  actions: ["s3:GetObject"],
  resources: ["databaseSecretArn"],
});

describe("The GuScheduledEcsTask pattern", () => {
  it("should use the specified container", () => {
    const stack = simpleGuStackForTesting();

    new GuScheduledEcsTask(stack, {
      schedule: Schedule.rate(Duration.minutes(1)),
      containerConfiguration: { id: "node:10", type: "registry" },
      monitoringConfiguration: { noMonitoring: true },
      vpc: vpc(stack),
      stack: "test",
      stage: "TEST",
      app: "ecs-test",
    });

    expect(stack).toHaveResourceLike("AWS::ECS::TaskDefinition", { ContainerDefinitions: [{ Image: "node:10" }] });
  });

  it("should use the specified schedule", () => {
    const stack = simpleGuStackForTesting();

    new GuScheduledEcsTask(stack, {
      schedule: Schedule.rate(Duration.minutes(1)),
      containerConfiguration: { id: "node:10", type: "registry" },
      monitoringConfiguration: { noMonitoring: true },
      vpc: vpc(stack),
      stack: "test",
      stage: "TEST",
      app: "ecs-test",
    });

    expect(stack).toHaveResourceLike("AWS::Events::Rule", { ScheduleExpression: "rate(1 minute)" });
  });

  it("should create the correct resources with lots of config", () => {
    const stack = simpleGuStackForTesting();

    new GuScheduledEcsTask(stack, {
      schedule: Schedule.rate(Duration.minutes(1)),
      containerConfiguration: { id: "node:10", type: "registry" },
      vpc: vpc(stack),
      stack: "test",
      stage: "TEST",
      app: "ecs-test",
      taskTimeoutInMinutes: 60,
      cpu: 1024,
      memory: 1024,
      monitoringConfiguration: { snsTopicArn: "arn:something:else:here:we:goalarm-topic", noMonitoring: false },
      taskCommand: `echo "yo ho row ho it's a pirates life for me"`,
      securityGroups: [securityGroup(stack)],
      customTaskPolicies: [testPolicy],
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
