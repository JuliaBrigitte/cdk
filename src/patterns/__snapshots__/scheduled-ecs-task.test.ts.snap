// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuScheduledEcsTask pattern should create the correct resources with lots of config 1`] = `
Object {
  "Outputs": Object {
    "StateMachineArnOutputEcstest": Object {
      "Value": Object {
        "Ref": "StateMachineEcstest84BB1B77",
      },
    },
  },
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ClusterEcstest1892FB90": Object {
      "Properties": Object {
        "ClusterName": "ecs-test-cluster-TEST",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::ECS::Cluster",
    },
    "ClusterEcstestAB072685": Object {
      "Properties": Object {
        "CapacityProviders": Array [
          "FARGATE",
          "FARGATE_SPOT",
        ],
        "Cluster": Object {
          "Ref": "ClusterEcstest1892FB90",
        },
        "DefaultCapacityProviderStrategy": Array [],
      },
      "Type": "AWS::ECS::ClusterCapacityProviderAssociations",
    },
    "ScheduleRuleEcstest07F5BBD2": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "StateMachineEcstest84BB1B77",
            },
            "Id": "Target0",
            "RoleArn": Object {
              "Fn::GetAtt": Array [
                "StateMachineEcstestEventsRole50FE033D",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "StateMachineEcstest84BB1B77": Object {
      "DependsOn": Array [
        "StateMachineEcstestRoleDefaultPolicy1EED542B",
        "StateMachineEcstestRoleAFD3E1C6",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"ecs-test-task\\",\\"States\\":{\\"ecs-test-task\\":{\\"End\\":true,\\"Type\\":\\"Task\\",\\"TimeoutSeconds\\":3600,\\"ResultPath\\":null,\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::ecs:runTask.sync\\",\\"Parameters\\":{\\"Cluster\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "ClusterEcstest1892FB90",
                  "Arn",
                ],
              },
              "\\",\\"TaskDefinition\\":\\"test-TEST-ecs-test\\",\\"NetworkConfiguration\\":{\\"AwsvpcConfiguration\\":{\\"Subnets\\":[\\"abc-123\\"],\\"SecurityGroups\\":[\\"id-123\\"]}},\\"LaunchType\\":\\"FARGATE\\",\\"PlatformVersion\\":\\"LATEST\\"}}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "StateMachineEcstestRoleAFD3E1C6",
            "Arn",
          ],
        },
        "StateMachineName": "ecs-test-TEST",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "StateMachineEcstestEventsRole50FE033D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEcstestEventsRoleDefaultPolicy2A914AF8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "StateMachineEcstest84BB1B77",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEcstestEventsRoleDefaultPolicy2A914AF8",
        "Roles": Array [
          Object {
            "Ref": "StateMachineEcstestEventsRole50FE033D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineEcstestRoleAFD3E1C6": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "states.",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEcstestRoleDefaultPolicy1EED542B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ecs:RunTask",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Fn::Select": Array [
                        1,
                        Object {
                          "Fn::Split": Array [
                            ":",
                            Object {
                              "Ref": "TaskDefinitionEcstest258379BD",
                            },
                          ],
                        },
                      ],
                    },
                    ":",
                    Object {
                      "Fn::Select": Array [
                        2,
                        Object {
                          "Fn::Split": Array [
                            ":",
                            Object {
                              "Ref": "TaskDefinitionEcstest258379BD",
                            },
                          ],
                        },
                      ],
                    },
                    ":",
                    Object {
                      "Fn::Select": Array [
                        3,
                        Object {
                          "Fn::Split": Array [
                            ":",
                            Object {
                              "Ref": "TaskDefinitionEcstest258379BD",
                            },
                          ],
                        },
                      ],
                    },
                    ":",
                    Object {
                      "Fn::Select": Array [
                        4,
                        Object {
                          "Fn::Split": Array [
                            ":",
                            Object {
                              "Ref": "TaskDefinitionEcstest258379BD",
                            },
                          ],
                        },
                      ],
                    },
                    ":",
                    Object {
                      "Fn::Select": Array [
                        0,
                        Object {
                          "Fn::Split": Array [
                            "/",
                            Object {
                              "Fn::Select": Array [
                                5,
                                Object {
                                  "Fn::Split": Array [
                                    ":",
                                    Object {
                                      "Ref": "TaskDefinitionEcstest258379BD",
                                    },
                                  ],
                                },
                              ],
                            },
                          ],
                        },
                      ],
                    },
                    "/",
                    Object {
                      "Fn::Select": Array [
                        1,
                        Object {
                          "Fn::Split": Array [
                            "/",
                            Object {
                              "Fn::Select": Array [
                                5,
                                Object {
                                  "Fn::Split": Array [
                                    ":",
                                    Object {
                                      "Ref": "TaskDefinitionEcstest258379BD",
                                    },
                                  ],
                                },
                              ],
                            },
                          ],
                        },
                      ],
                    },
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ecs:StopTask",
                "ecs:DescribeTasks",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TaskDefinitionEcstestTaskRoleD5169AE2",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "TaskDefinitionEcstestExecutionRole9D2A7D47",
                    "Arn",
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "events:PutTargets",
                "events:PutRule",
                "events:DescribeRule",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":events:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":rule/StepFunctionsGetEventsForECSTaskRule",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEcstestRoleDefaultPolicy1EED542B",
        "Roles": Array [
          Object {
            "Ref": "StateMachineEcstestRoleAFD3E1C6",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskDefinitionEcstest258379BD": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Command": Array [
              "-c",
              "echo \\"yo ho row ho it's a pirates life for me\\"",
            ],
            "Cpu": 1024,
            "EntryPoint": Array [
              "/bin/sh",
            ],
            "Essential": true,
            "Image": "node:10",
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "TaskDefinitionEcstestScheduledTaskContainerEcstestLogGroup1508FA51",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "ecs-test",
              },
            },
            "Memory": 1024,
            "Name": "ScheduledTaskContainerEcstest",
          },
        ],
        "Cpu": "1024",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionEcstestExecutionRole9D2A7D47",
            "Arn",
          ],
        },
        "Family": "test-TEST-ecs-test",
        "Memory": "1024",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionEcstestTaskRoleD5169AE2",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionEcstestExecutionRole9D2A7D47": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskDefinitionEcstestExecutionRoleDefaultPolicy562096CB": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TaskDefinitionEcstestScheduledTaskContainerEcstestLogGroup1508FA51",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskDefinitionEcstestExecutionRoleDefaultPolicy562096CB",
        "Roles": Array [
          Object {
            "Ref": "TaskDefinitionEcstestExecutionRole9D2A7D47",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskDefinitionEcstestScheduledTaskContainerEcstestLogGroup1508FA51": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TaskDefinitionEcstestTaskRoleD5169AE2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "ecs-test",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskDefinitionEcstestTaskRoleDefaultPolicyC6A4D178": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:s3:::",
                    Object {
                      "Ref": "DistributionBucketName",
                    },
                    "/test-stack/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/ecs-test/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": "databaseSecretArn",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskDefinitionEcstestTaskRoleDefaultPolicyC6A4D178",
        "Roles": Array [
          Object {
            "Ref": "TaskDefinitionEcstestTaskRoleD5169AE2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ecstestexecutionfailedC93F511B": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          "arn:something:else:here:we:goalarm-topic",
        ],
        "AlarmDescription": "ecs-test-TEST job failed ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StateMachineArn",
            "Value": Object {
              "Ref": "StateMachineEcstest84BB1B77",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsFailed",
        "Namespace": "AWS/States",
        "Period": 3600,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ecstesttimeoutFE335653": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          "arn:something:else:here:we:goalarm-topic",
        ],
        "AlarmDescription": "ecs-test-TEST job timed out ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "StateMachineArn",
            "Value": Object {
              "Ref": "StateMachineEcstest84BB1B77",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsTimedOut",
        "Namespace": "AWS/States",
        "Period": 3600,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
